### PR TITLE
[MODFISTO-290] - Allowing "Available" to be a negative number

### PR DESCRIPTION
**Pull request by azizbekxm**
## Purpose

https://issues.folio.org/browse/MODFISTO-290
## Approach

- removed [CalculationUtils.java#L55](https://github.com/folio-org/mod-finance-storage/blob/a18a4335fd1f33285dbcee1ad591e12f3198158b/src/main/java/org/folio/utils/CalculationUtils.java#L55).
- updated unit tests
- updated karate tests
## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?
- Did you modify code to call some additional endpoints?
  - [ ] If so, do you check that necessary module permission added in ModuleDescriptor-template.yaml file?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by azizbekxm** [comment]
image

File: src/main/java/org/folio/service/budget/BudgetService.java [hunk 1]
```diff
@@ -219,23 +218,14 @@ public void clearReadOnlyFields(Budget budgetFromNew) {
   }
 
   public Future<Void> checkBudgetHaveMoneyForTransaction(Transaction transaction, DBClient client) {
-    if (transaction.getFromFundId() == null) {
+    if (transaction.getFromFundId() == null || transaction.getTransactionType() == Transaction.TransactionType.TRANSFER) {
```
Comment (Abdulkhakimov) [inline src/main/java/org/folio/service/budget/BudgetService.java:221]: To improve readability please replace with java.util.Objects.isNull static method this statement : "transaction.getFromFundId() == null"

File: src/test/java/org/folio/rest/impl/TransactionTest.java [hunk 2]
```diff
@@ -274,8 +273,9 @@ void testCreateTransferThatDoesNotHaveEnoughMoney() throws MalformedURLException
 
     // create Transfer
     postData(TRANSACTION_ENDPOINT, transactionSample, TRANSACTION_TENANT_HEADER).then()
-      .statusCode(400)
-      .body(containsString(TRANSFER_NOT_ENOUGH_MONEY_ERROR_TEXT));
+      .statusCode(201)
```
Comment (Abdulkhakimov) [inline src/test/java/org/folio/rest/impl/TransactionTest.java:276]: Need to assert negative available ammount from response payload

**Review by SinghAdes — COMMENTED** [review]
Please check "BaseTransactionRestrictionService.java" also , i believe there are some validations for negative amounts , which needs to be updated.

File: src/main/java/org/folio/service/budget/BudgetService.java (hunk 1 above)
Comment (azizbekxm) [reply src/main/java/org/folio/service/budget/BudgetService.java:221]: Ok, I will fix, thanks

File: src/test/java/org/folio/rest/impl/TransactionTest.java (hunk 2 above)
Comment (azizbekxm) [reply src/test/java/org/folio/rest/impl/TransactionTest.java:276]: ok, I will add these cases also

**Comment by Abdulkhakimov** [comment]
> Please check "BaseTransactionRestrictionService.java" also , i believe there are some validations for negative amounts , which needs to be updated.

Could you tell me what you mean exactly? Thank you!

**Comment by azizbekxm** [comment]
> > Please check "BaseTransactionRestrictionService.java" also , i believe there are some validations for negative amounts , which needs to be updated.
> 
> Could you tell me what you mean exactly? Thank you!
A month ago Adesh worked this similar task maybe Adesh want to say this diff
https://github.com/folio-org/mod-finance-storage/pull/320/files#diff-b74f200ab893091080bdc7e7da9898f6d2b11ef7cd6ec8db90d36534494f3f7f

**Comment by SinghAdes** [comment]
> > Please check "BaseTransactionRestrictionService.java" also , i believe there are some validations for negative amounts , which needs to be updated.
> 
> Could you tell me what you mean exactly? Thank you!

I was talking about this 
https://github.com/folio-org/mod-finance-storage/blob/master/src/main/java/org/folio/service/transactions/restriction/BaseTransactionRestrictionService.java#:~:text=public%20Future%3C,%7D)%3B

**Comment by SinghAdes** [comment]
> > > Please check "BaseTransactionRestrictionService.java" also , i believe there are some validations for negative amounts , which needs to be updated.
> > 
> > 
> > Could you tell me what you mean exactly? Thank you!
> 
> I was talking about this https://github.com/folio-org/mod-finance-storage/blob/master/src/main/java/org/folio/service/transactions/restriction/BaseTransactionRestrictionService.java#:~:text=public%20Future%3C,%7D)%3B





Here is the PR..https://github.com/folio-org/mod-finance-storage/pull/320

**Comment by azizbekxm** [comment]
> > > > Please check "BaseTransactionRestrictionService.java" also , i believe there are some validations for negative amounts , which needs to be updated.
> > > 
> > > 
> > > Could you tell me what you mean exactly? Thank you!
> > 
> > 
> > I was talking about this https://github.com/folio-org/mod-finance-storage/blob/master/src/main/java/org/folio/service/transactions/restriction/BaseTransactionRestrictionService.java#:~:text=public%20Future%3C,%7D)%3B
> 
> Here is the PR..#320

**Comment by Abdulkhakimov** [comment]
> > > > Please check "BaseTransactionRestrictionService.java" also , i believe there are some validations for negative amounts , which needs to be updated.
> > > 
> > > 
> > > Could you tell me what you mean exactly? Thank you!
> > 
> > 
> > I was talking about this https://github.com/folio-org/mod-finance-storage/blob/master/src/main/java/org/folio/service/transactions/restriction/BaseTransactionRestrictionService.java#:~:text=public%20Future%3C,%7D)%3B
> 
> Here is the PR..#320

During the spike, I also looked at your PR and came to the conclusion that we don't need to change these places to avoid problems with certain restrictions for encumbrance, pending payment and payment.

**Comment by Abdulkhakimov** [comment]
@SinghAdes Dennis Bridges also propose to leave these retrictions : https://issues.folio.org/browse/MODFISTO-361?focusedCommentId=152757&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-152757

**Comment by SinghAdes** [comment]
> @SinghAdes Dennis Bridges also propose to leave these retrictions : https://issues.folio.org/browse/MODFISTO-361?focusedCommentId=152757&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-152757

okay , Thanks for clarification.

File: src/test/java/org/folio/rest/impl/TransactionTest.java [hunk 3]
```diff
@@ -44,7 +43,7 @@ public class TransactionTest extends TestBase {
 
   private static final String FY_FUND_QUERY = "?query=fiscalYearId==%s AND fundId==%s";
   public static final String ALLOCATION_SAMPLE = "data/transactions/zallocation_AFRICAHIST-FY22_ANZHIST-FY22.json";
-  public static final String TRANSFER_NOT_ENOUGH_MONEY_ERROR_TEXT = "Transfer was not successful. There is not enough money Available in the budget to complete this Transfer";
+//  public static final String TRANSFER_NOT_ENOUGH_MONEY_ERROR_TEXT = "Transfer was not successful. There is not enough money Available in the budget to complete this Transfer";
```
Comment (SerhiiNosko) [inline src/test/java/org/folio/rest/impl/TransactionTest.java:46]: If this code not used any more - please delete it

Comment (azizbekxm) [reply src/test/java/org/folio/rest/impl/TransactionTest.java:46]: ok, I will delete

**Review by Abdulkhakimov — APPROVED** [review]

**Comment by azizbekxm** [comment]
image

image

**Review by SerhiiNosko — APPROVED** [review]

**Review by SinghAdes — APPROVED** [review]

**Pull request merged**